libraries/chain/include/eosio/chain/backing_store/kv_context_rocksdb.hpp

```diff
@@ -0,0 +1,451 @@
+                     account_name payer) override {
+         EOS_ASSERT(name{ contract } == receiver, table_operation_not_permitted, "Can not write to this key");
+         EOS_ASSERT(key_size <= limits.max_key_size, kv_limit_exceeded, "Key too large");
+         EOS_ASSERT(value_size <= limits.max_value_size, kv_limit_exceeded, "Value too large");
+
+         current_value.reset();
+         current_key                = eosio::session::shared_bytes{};
+         auto old_value             = std::optional<eosio::session::shared_bytes>{};
+         auto old_value_size = int64_t{ 0 };
+         try {
+            try {
+               auto composite_key = make_composite_key(contract, nullptr, 0, key, key_size);
+               old_value          = session->read(composite_key);
+               if (old_value) {
+                  old_value_size = backing_store::actual_value_size(old_value->size());
+               }
+
+               auto new_value = build_value(value, value_size, payer);
+               session->write(composite_key, new_value);
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
+
+         auto resource_delta = int64_t{ 0 };
+         if (old_value) {
+            account_name old_payer = backing_store::get_payer(old_value->data());
+            resource_delta =
+                  update_table_usage(resource_manager, old_payer, payer, key, key_size, old_value_size, value_size);
+         } else {
+            resource_delta = create_table_usage(resource_manager, payer, key, key_size, value_size);
+         }
+
+         return resource_delta;
+      }
+
+      bool kv_get(uint64_t contract, const char* key, uint32_t key_size, uint32_t& value_size) override {
+         current_key = eosio::session::shared_bytes{};
+         current_value.reset();
+         try {
+            try {
+               current_key   = make_composite_key(contract, nullptr, 0, key, key_size);
+               current_value = session->read(current_key);
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
+
+         if (current_value) {
+            value_size = backing_store::actual_value_size(current_value->size());
+            return true;
+         } else {
+            value_size = 0;
+            return false;
+         }
+      }
+
+      uint32_t kv_get_data(uint32_t offset, char* data, uint32_t data_size) override {
+         const char* temp        = nullptr;
+         uint32_t      temp_size = 0;
+         if (current_value) {
+            temp      = backing_store::actual_value_start(current_value->data());
+            temp_size = backing_store::actual_value_size(current_value->size());
+         }
+         if (offset < temp_size)
+            memcpy(data, temp + offset, std::min(data_size, temp_size - offset));
+         return temp_size;
```

Comment:
The docs says "@return number of bytes written in data." However, this does not return the number of bytes written to `data` in case where `offset >= temp_size` or when `data_size` is less than `temp_size - offset`. This goes for the kv_context_chainbase impl as well.

The test cases for this seem to have been written for what was implemented not for what it should do.

Reply:
The document https://github.com/EOSIO/spec-repo/blob/master/esr_key_value_database_intrinsics.md says
"Fetches data from temporary buffer starting at offset. Copies up to data_size bytes into data. Returns amount of data in temporary buffer.". Which document are you referring to?

Reply:
`libraries/chain/include/eosio/chain/webassembly/interface.hpp` which is what the documentation is generated from I believe.

Reply:
https://github.com/EOSIO/spec-repo/blob/master/esr_key_value_database_intrinsics.md says "It still aborts the transaction if [data, data + data_size) is out of the WASM's linear memory range." which is not true of the implementation.

Reply:
This is handled because of the fact that we are only using eos-vm backends now.  There is no need to explicitly force any of that as it will segfault all of the backends.

Reply:
Is the consensus here that the comment in `interface.hpp` needs to be updated? Under the current impl how do you know if it failed?

Reply:
Should we  change the comment in `libraries/chain/include/eosio/chain/webassembly/interface.hpp` to make it consistent with the implementation and https://github.com/EOSIO/spec-repo/blob/master/esr_key_value_database_intrinsics.md?

Reply:
This is a doc problem.  @nksanthosh  who is the doc person, I cannot think of his name right now?

Reply:
EPE-461 now is in charge of this

Reply:
Under the current impl how do you know if `kv_get_data` failed? 

Say you pass in too large of an offset. `kv_get_data` will return the size of the temp buffer but not copy anything into `data`.

Reply:
Should we throw an exception in this case? I'd also like to assert on `data` not being NULL.